File: brainfuck/brainfuck.py
import platform
import sys
import socket
import os
from pathlib import Path

class Tape(object):
    def __init__(self):
        self.tape = [0]
        self.pos = 0
    def get(self):
        return self.tape[self.pos]
    def inc(self):
        self.tape[self.pos] += 1
    def dec(self):
        self.tape[self.pos] -= 1
    def advance(self):
        self.pos += 1
        if len(self.tape) <= self.pos:
            self.tape.append(0)
    def devance(self):
        if self.pos > 0:
            self.pos -= 1

class Program(object):
    def __init__(self, text):
        self.code = ""
        self.bracket_map = {}
        leftstack = []

        pc = 0
        for char in text:
            if char in ('[', ']', '<', '>', '+', '-', ',', '.'):
                self.code += char
                if char == '[':
                    leftstack.append(pc)
                elif char == ']' and len(leftstack) > 0:
                    left = leftstack.pop()
                    right = pc
                    self.bracket_map[left] = right
                    self.bracket_map[right] = left
                pc += 1

    def run(self):
        pc = 0
        tape = Tape()

        while pc < len(self.code):
            char = self.code[pc]
            if char == "+":
                tape.inc()
            elif char == "-":
                tape.dec()
            elif char == ">":
                tape.advance()
            elif char == "<":
                tape.devance()
            elif char == "[" and tape.get() == 0:
                pc = self.bracket_map[pc]
            elif char == "]" and tape.get() != 0:
                pc = self.bracket_map[pc]
            elif char == ".":
                sys.stdout.write(chr(tape.get()))
                sys.stdout.flush()
            pc += 1

def notify(msg):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        if not s.connect_ex(("localhost", 9001)):
            s.sendall(bytes(msg, 'utf8'))

text = Path(sys.argv[1]).read_text()

notify("%s\t%d" % (platform.python_implementation(), os.getpid()))

Program(text).run()

notify("stop")

File: brainfuck/brainfuck.js
var Tape = function() {
	var pos = 0, tape = [0];

	this.inc = function() { tape[pos]++; }
	this.dec = function() { tape[pos]--; }
	this.advance = function() { pos++; if (tape.length <= pos) tape.push(0); }
	this.devance = function() { if (pos > 0) pos--; }
	this.get = function() { return tape[pos]; }
}

var Brainfuck = function(text) {
	var me = this;
	me.code = "";
	me.bracket_map = function(text) {
		var leftstack = [];
		var bm = {};

		for (var i = 0, pc = 0; i < text.length; i++) {
			var c = text.charAt(i);
			if ("+-<>[].,".indexOf(c) === -1) continue;
			if (c === '[') leftstack.push(pc);
			if (c === ']' && leftstack.length > 0) {
				var left = leftstack.pop();
				bm[left] = pc;
				bm[pc] = left;
 			}
 			me.code += c;
 			pc++;
		}
		return bm;
	}(text);

	me.run = function() {
		var pc = 0;
		var tape = new Tape();
		var code = this.code;
		var bm = this.bracket_map;

		for (var pc = 0; pc < code.length; pc++)
			switch(code[pc]) {
				case '+': tape.inc(); break;
				case '-': tape.dec(); break;
				case '>': tape.advance(); break;
				case '<': tape.devance(); break;
				case '[': if (tape.get() == 0) pc = bm[pc]; break;
				case ']': if (tape.get() != 0) pc = bm[pc]; break;
				case '.': process.stdout.write(String.fromCharCode(tape.get()));
				default:
			}
	};

}

var text = require('fs').readFileSync(process.argv[2].toString()).toString(); 
var brainfuck = new Brainfuck(text);
brainfuck.run();
